pkg/kubelet/device-plugin/endpoint.go

```diff
@@ -0,0 +1,172 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package deviceplugin
+
+import (
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+
+	pluginapi "k8s.io/kubernetes/pkg/kubelet/apis/device-plugin/v1alpha1"
+)
+
+type endpoint struct {
+	client pluginapi.DevicePluginClient
+
+	socketPath   string
+	resourceName string
+
+	devices []*pluginapi.Device
+	mutex   sync.Mutex
+
+	callback MonitorCallback
+
+	cancel context.CancelFunc
+	ctx    context.Context
+}
+
+func newEndpoint(socketPath, resourceName string,
+	callback MonitorCallback) (*endpoint, error) {
+
+	client, err := dial(socketPath)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, stop := context.WithCancel(context.Background())
+
+	return &endpoint{
+		client: client,
+
+		socketPath:   socketPath,
+		resourceName: resourceName,
+
+		devices:  nil,
+		callback: callback,
+
+		cancel: stop,
+		ctx:    ctx,
+	}, nil
+}
+
+func (e *endpoint) List() (pluginapi.DevicePlugin_ListAndWatchClient, error) {
+
+	stream, err := e.client.ListAndWatch(e.ctx, &pluginapi.Empty{})
+	if err != nil {
+		glog.Errorf("Could not call ListWatch for device plugin %s: %v",
+			e.resourceName, err)
+
+		return nil, err
+	}
+
+	devs, err := stream.Recv()
+	if err != nil {
+		glog.Errorf(ErrListWatch, e.resourceName, err)
+		return nil, err
+	}
+
+	e.mutex.Lock()
+	e.devices = devs.Devices
+	e.mutex.Unlock()
+
+	return stream, nil
+}
+
+func (e *endpoint) ListWatch(stream pluginapi.DevicePlugin_ListAndWatchClient) {
+	glog.V(2).Infof("Starting ListWatch")
+
+	for {
+		response, err := stream.Recv()
+		if err != nil {
+			glog.Errorf(ErrListWatch, e.resourceName, err)
+			return
+		}
+
+		devs := response.Devices
+		glog.V(2).Infof("State pushed for device plugin %s", e.resourceName, err)
+
+		// TODO check this is the same list
+		if len(devs) > len(e.devices) {
+			glog.Errorf("New device for Device Plugin %s", e.resourceName)
+			continue
+		}
+
+		var updated []*pluginapi.Device
+
+		e.mutex.Lock()
+		for i := 0; i < len(e.devices); i++ {
+			d1 := e.devices[i]
+			d2, ok := GetDevice(d1, devs)
+
+			// TODO Should we remove it ?
+			if !ok {
+				glog.Errorf("Device lost for Device Plugin %s", e.resourceName)
+				continue
+			}
+
+			if d2.Health == d1.Health {
+				continue
+			}
+
+			glog.V(5).Infof("Health changed from %s to %s for device %s",
+				d1.Health, d2.Health, e.resourceName)
+
```

Comment:
We may consider to add an error log when a device state changes from health to unhealth for easy monitoring through node problem detector.